agent/engine/docker_task_engine.go

```diff
@@ -625,7 +647,21 @@ func (engine *DockerTaskEngine) startContainer(task *api.Task, container *api.Co
 			Error: CannotStartContainerError{fmt.Errorf("Container not recorded as created")},
 		}
 	}
-	return client.StartContainer(dockerContainer.DockerID, startContainerTimeout)
+	dockerContainerMD := client.StartContainer(dockerContainer.DockerID, startContainerTimeout)
+
+	// Get metadata through container inspection and available task information then write this to the metadata file
+	// Performs this in the background to avoid delaying container start
```

Comment:
It almost feels like this should be a new container state.  As is, if the agent restarts while this goroutine is running (after the container has started and we've recorded that in the state file, but before we've completed updating the metadata), you may never get metadata properly updated (`startContainer` won't be called again).